products/marketing_analytics/backend/hogql_queries/__snapshots__/test_conversion_goal_processor.ambr

```diff
@@ -1,17 +1,236 @@
 # serializer version: 1
+# name: TestConversionGoalProcessor.test_attribution_window_sql_structure_demo
+  '''
+  
+  SELECT
+      if(notEmpty(campaign_name), campaign_name, 'organic') AS campaign,
+      if(notEmpty(source_name), source_name, 'organic') AS source,
+      count() AS conversion_0
+  
+  FROM
+      (
+  SELECT
+          person_id,
+          if(notEmpty(conversion_campaign), conversion_campaign, if(notEmpty(fallback_campaign), fallback_campaign, '')) AS campaign_name,
+          if(notEmpty(conversion_source), conversion_source, if(notEmpty(fallback_source), fallback_source, '')) AS source_name,
+          1 AS conversion_value
+      
+  FROM
+          (
+  SELECT
+              person_id,
+              conversion_timestamps[i] AS conversion_time,
+              conversion_math_values[i] AS conversion_math_value,
+              conversion_campaigns[i] AS conversion_campaign,
+              conversion_sources[i] AS conversion_source,
+              arrayMax(arrayFilter(x -> and(lessOrEquals(x, conversion_timestamps[i]), greaterOrEquals(x, minus(conversion_timestamps[i], 15552000))), utm_timestamps)) AS last_utm_timestamp,
```

Comment:
Did you try this with a window function approach like you did with the session duration ticket? I am not sure if it will work or be faster for your use case, but noting it so you can check.

```
lagInFrame(properties.utm_campaign) OVER (
          PARTITION BY person_id
          ORDER BY timestamp
          ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW
      ) AS last_utm_campaign
```